### PR TITLE
Adding threadpool metrics

### PR DESCRIPTION
**Pull request by gareth-ellis**
This implements metrics for the threadpools.

The aim is to emit metrics for the various threadpools, the metric callback should be created when the threadpool is created, and removed before the threadpool is shutdown.

The PR also includes a test for the new metrics, and some additions to the metrics test plugin.

Finally the metric name check has been modified to allow some of the non compliant threadpools (too long, includes - )

### CODE REVIEW

**Review by pgomulka — COMMENTED** [review]
it looks good, I left few comments/questions
btw do you think it is worth adding an integration test for this?
something along the lines of https://github.com/elastic/elasticsearch/blob/main/modules/repository-s3/src/internalClusterTest/java/org/elasticsearch/repositories/s3/S3BlobStoreRepositoryTests.java#L229
it would have to be an integration test currently - I think we should have something for lower level unit testing though, I will create an issue for this

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsRejectedExecutionHandler.java [hunk 1]
```diff
@@ -9,13 +9,16 @@
...
+import org.elasticsearch.telemetry.metric.DoubleCounter;
+import org.elasticsearch.telemetry.metric.MeterRegistry;
 
 import java.util.concurrent.RejectedExecutionHandler;
 import java.util.concurrent.ThreadPoolExecutor;
 
 public abstract class EsRejectedExecutionHandler implements RejectedExecutionHandler {
 
     private final CounterMetric rejected = new CounterMetric();
+    private DoubleCounter telemetry_rejected = null;
```
[the remark this answers is not in the document]
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsRejectedExecutionHandler.java:21]: ++

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 2]
```diff
@@ -321,6 +328,64 @@ public ThreadPool(final Settings settings, final ExecutorBuilder<?>... customBui
...
 
+    private static HashMap<String, LongGauge> setupMetrics(MeterRegistry meterRegistry, String name, ExecutorHolder holder) {
+        Map<String, Object> at = null;
+        HashMap<String, LongGauge> gauges = new HashMap<>();
+        if (holder.executor() instanceof ThreadPoolExecutor threadPoolExecutor) {
+            String prefix = "es.tp." + name + ".t."; // temporarily shortening due to max 68 chars limit
+            gauges.put(
+                "threads",
+                meterRegistry.registerLongGauge(
+                    prefix + "threads",
```
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:339]: E.g. here, would it `es.thread_pool.<name>.thread.threads`? if so, I'd suggest `es.thread_pool.<name>.thread.count`

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 3]
```diff
@@ -321,6 +328,64 @@ public ThreadPool(final Settings settings, final ExecutorBuilder<?>... customBui
         this.cachedTimeThread.start();
     }
 
+    private static HashMap<String, LongGauge> setupMetrics(MeterRegistry meterRegistry, String name, ExecutorHolder holder) {
+        Map<String, Object> at = null;
+        HashMap<String, LongGauge> gauges = new HashMap<>();
```
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:333]: It's not clear to me why this is a map - we are never using the keys, only the values. Maybe a List will be enough?

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 4]
```diff
@@ -553,6 +628,16 @@ public void shutdownNow() {
         scheduler.shutdownNow();
         for (ExecutorHolder executor : executors.values()) {
             if (executor.executor() instanceof ThreadPoolExecutor) {
+                if (this.gauges.containsKey(executor.info.getName())) {
```
[the remark this answers is not in the document]
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:631]: ++

**Review by rjernst — COMMENTED** [review]
I have a few thoughts

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 5]
```diff
@@ -321,6 +328,64 @@ public ThreadPool(final Settings settings, final ExecutorBuilder<?>... customBui
         this.cachedTimeThread.start();
     }
 
+    private static HashMap<String, LongGauge> setupMetrics(MeterRegistry meterRegistry, String name, ExecutorHolder holder) {
+        Map<String, Object> at = null;
+        HashMap<String, LongGauge> gauges = new HashMap<>();
+        if (holder.executor() instanceof ThreadPoolExecutor threadPoolExecutor) {
+            String prefix = "es.tp." + name + ".t."; // temporarily shortening due to max 68 chars limit
```
[the remark this answers is not in the document]
Comment (gareth-ellis) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:335]: Yes, we hit the limit already - the full name i was thinking was es.thread_pool.<name>.thread.<metric>.
(i see the limit was 63, not 68 like I wrote)
Some of the threadpools have fairly long names. E.g » 
[2023-11-28T10:23:04,176][ERROR][o.e.b.Elasticsearch      ] [runTask-0] fatal exception while booting Elasticsearch java.lang.IllegalArgumentException: Instrument name [es.thread_pool.searchable_snapshots_cache_fetch_async.threads.count] with length [67] exceeds maximum length [63]

I know it will be disruptive to change afterwards, I am not sure what sort of timeline there is for the new apm agent, if thats not too far away we could perhaps wait on merging until that's available.

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java (hunk 2 above)
Comment (gareth-ellis) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:339]: Yes, ok - I'll change to count instead of threads

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:335]: I should have this fixed soon https://github.com/elastic/elasticsearch/pull/102691

[the remark this answers is not in the document]
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:335]: https://github.com/elastic/elasticsearch/pull/102691 is now merged, @gareth-ellis you can update the names and use up to 255 characters :)

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java (hunk 3 above)
Comment (gareth-ellis) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: Changed to an ArrayList

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java (hunk 2 above)
Comment (gareth-ellis) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: Fixed in fc45e177b705d1b49fff5039f1cafd714c267d97

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (gareth-ellis) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: Fixed now in  fc45e177b705d1b49fff5039f1cafd714c267d97

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (gareth-ellis) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: Fixed in fc45e177b705d1b49fff5039f1cafd714c267d97

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 6]
```diff
@@ -543,6 +608,16 @@ public void shutdown() {
         scheduler.shutdown();
         for (ExecutorHolder executor : executors.values()) {
             if (executor.executor() instanceof ThreadPoolExecutor) {
+                if (this.gauges.containsKey(executor.info.getName())) {
+                    this.gauges.get(executor.info.getName()).forEach((k, v) -> {
+                        try {
+                            v.close();
+                        } catch (Exception e) {
+                            logger.warn(format("Failed to close LongGauge for %s. %s", executor.info.getName(), e.getMessage()));
```
[the remark this answers is not in the document]
Comment (gareth-ellis) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: Added exception to final parameter of warn message in 01260a3599eef59426d505b47d4193d0e27b3339

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 7]
```diff
@@ -321,6 +328,64 @@ public ThreadPool(final Settings settings, final ExecutorBuilder<?>... customBui
...
+                meterRegistry.registerLongGauge(
+                    prefix + "largest",
+                    "largest pool size for " + name,
+                    "count",
+                    () -> new LongWithAttributes(threadPoolExecutor.getLargestPoolSize(), at)
+                )
+            );
+            gauges.put(
+                "completed",
+                meterRegistry.registerLongGauge(
```
[the remark this answers is not in the document]
Comment (gareth-ellis) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: I've swapped to a Counter. 
I would think a lot of the metrics related to threadpools are mostly in relation to the lifetime of a node, so on restart all metrics will return to 0, and at which point a new node that might get started would end up in a new pod, which I would expect is what we will group a lot of the metrics by

**Comment by piergm** [comment]
Hi @gareth-ellis, I wonder if all the thread-pool related metrics could be instrumented in NodeMetrics.java (see other `_node/stats` metrics instrumented here #102248) if you'd like I can make a draft PR to showcase the implementation I have in mind.

Another note I'd like to make is that as of right now we have 41 different thread-pools each with 6 stats (active, queue, rejected, ...). A solution I can envision is to have only the 6 stats and have the thread-pool name as attribute to the stats. I am not sure this is a viable solution but I wanted your input on it. Furthermore, I think that the implementation of this particular solution would be tricky since my understanding is that there is no way to have the callback returning multiple LongWithAttributes. I do have a way around it but it feels too hacky...

What are your opinions on these topics?

**Comment by gareth-ellis** [comment]
Hi @piergm ,

I had a look at your implementation of NodeMetrics and it probably should work for threadpools too - Is there a benefit though? We would essentially be running the same code, the only difference I guess would be the collection would be async from apm server collecting the stats  - the result though there would be we were collecting once a minute but (currently) only sending once every two minutes (since if I understand you fill the cache every minute, but APM will only poll for metrics every 120 seconds).

Regardinng the second point, I am also a little concerned with the number of metrics this will add, I don't know if the multiple attribute approach is any more or less favourable, the payload thats sent to APM is a single message anyway, rather than 1 message per metric - do you know about that?

**Comment by piergm** [comment]
Hi @gareth-ellis, I think the main benefit is the async collection as you said but also simpler/cleaner solution overall. Handing Metrics lifecycle in `ThreadPool` feels odd. 
Regarding your concern:
> the result though there would be we were collecting once a minute but (currently) only sending once every two minutes (since if I understand you fill the cache every minute, but APM will only poll for metrics every 120 seconds).

This is incorrect, `SingleObjectCache` do expire after one minutes but it's not refreshed until is needed. Therefore in our case the cache is refreshed (and stats collected) once every 120 seconds. This will work even if in the future we change the APM poll rate (unless we set it to less than 1 min). 
Basically the way it works rn in NodeMetrics is that every time APM starts polling the first callback will collect and cache stats while all the other metrics callbacks will have the stats cached. I set the expiry to 1 min to be less then polling rate but enough for all the stats to be collected.

> Regardinng the second point, I am also a little concerned with the number of metrics this will add, I don't know if the multiple attribute approach is any more or less favourable, the payload thats sent to APM is a single message anyway, rather than 1 message per metric - do you know about that?

I am not sure either what is the best solution in this case, @pgomulka WDYT?

**Comment by pgomulka** [comment]
I personally would prefer the metrics to be registered/used around the code that they measure. Unless it is really hard to implement. 
@piergm are you suggesting to move https://github.com/elastic/elasticsearch/pull/102371#diff-2835e151b1e7435ff4760a9263e0ecbc359597a9ff02005fc7fec099d38549e6R333 to NodeMetrics.java ?
I think what we have here now is simple, not sure we need to change this. Unless I am missing something

**Comment by piergm** [comment]
@pgomulka thank you for reviewing, I see your point and I agree It's fine to keep this as is.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsRejectedExecutionHandler.java [hunk 8]
```diff
@@ -26,6 +29,14 @@ public long rejected() {
 
     protected void incrementRejections() {
         rejected.inc();
+        if (telemetry_rejected != null) {
+            telemetry_rejected.increment();
+        }
+    }
+
+    public void registerCounter(MeterRegistry meterRegistry, String prefix, String name) {
+        telemetry_rejected = meterRegistry.registerDoubleCounter(prefix + "rejected", "number of rejected threads for " + name, "count");
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/EsRejectedExecutionHandler.java:38]: can we use 'total' suffix here?

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 9]
```diff
@@ -321,6 +330,60 @@ public ThreadPool(final Settings settings, final ExecutorBuilder<?>... customBui
...
 
+    private static ArrayList<Instrument> setupMetrics(MeterRegistry meterRegistry, String name, ExecutorHolder holder) {
+        Map<String, Object> at = Map.of();
+        ArrayList<Instrument> instruments = new ArrayList<>();
+        if (holder.executor() instanceof ThreadPoolExecutor threadPoolExecutor) {
+            String prefix = "es.thread_pool." + name + ".threads.";
+            logger.info("Starting tp metrics {}", name);
+            instruments.add(
+                meterRegistry.registerLongGauge(
+                    prefix + "count",
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:341]: can we use a `current` suffix here?

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 10]
```diff
@@ -321,6 +330,60 @@ public ThreadPool(final Settings settings, final ExecutorBuilder<?>... customBui
...
+                    "number of threads for " + name,
+                    "count",
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:343]: I know this feels odd to have a mandatory unit here, let's keep "count" here, we will look into this later

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 11]
```diff
@@ -321,6 +330,60 @@ public ThreadPool(final Settings settings, final ExecutorBuilder<?>... customBui
...
+                    () -> new LongWithAttributes(threadPoolExecutor.getPoolSize(), at)
+                )
+            );
+            instruments.add(
+                meterRegistry.registerLongGauge(
+                    prefix + "queue",
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:349]: can we rename to "queue.size" ?

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 12]
```diff
@@ -321,6 +330,60 @@ public ThreadPool(final Settings settings, final ExecutorBuilder<?>... customBui
...
+                    "number queue size for " + name,
+                    "count",
+                    () -> new LongWithAttributes(threadPoolExecutor.getQueue().size(), at)
+                )
+            );
+            instruments.add(
+                meterRegistry.registerLongGauge(
+                    prefix + "active",
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:357]: can we rename to "active.current"

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 13]
```diff
@@ -321,6 +330,60 @@ public ThreadPool(final Settings settings, final ExecutorBuilder<?>... customBui
...
+                    "number of active threads for " + name,
+                    "count",
+                    () -> new LongWithAttributes(threadPoolExecutor.getActiveCount(), at)
+                )
+            );
+            instruments.add(
+                meterRegistry.registerLongGauge(
+                    prefix + "largest",
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:365]: can we rename to "largest.current"

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 14]
```diff
@@ -321,6 +330,60 @@ public ThreadPool(final Settings settings, final ExecutorBuilder<?>... customBui
...
+                    "largest pool size for " + name,
+                    "count",
+                    () -> new LongWithAttributes(threadPoolExecutor.getLargestPoolSize(), at)
+                )
+            );
+            instruments.add(
+                meterRegistry.registerLongAsyncCounter(
+                    prefix + "completed",
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:373]: can we rename to "completed.total" ?

**Review by pgomulka — COMMENTED** [review]
I left some questions around naming

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsRejectedExecutionHandler.java (hunk 8 above)
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsRejectedExecutionHandler.java:38]: btw why double counter not a long counter?

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java (hunk 10 above)
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:343]: An alternative I've seen applied elsewhere: use the object name, e.g. `{thread}`. See https://opentelemetry.io/docs/specs/semconv/general/metrics/#instrument-units
Just a suggestion, we do not have guidelines nor mandate anything.

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java (hunk 14 above)
Comment (gareth-ellis) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: Addressed in 1a0b1defba83ade4b673cf9af21fa09161397f50

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java (hunk 12 above)
Comment (gareth-ellis) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: Addressed in 1a0b1defba83ade4b673cf9af21fa09161397f50

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java (hunk 11 above)
Comment (gareth-ellis) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: Addressed in 1a0b1defba83ade4b673cf9af21fa09161397f50

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java (hunk 9 above)
Comment (gareth-ellis) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: Addressed in 1a0b1defba83ade4b673cf9af21fa09161397f50

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsRejectedExecutionHandler.java (hunk 8 above)
Comment (gareth-ellis) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsRejectedExecutionHandler.java]: I am not too sure why I took double instead of long there, changed to Long now in 1a0b1defba83ade4b673cf9af21fa09161397f50

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsRejectedExecutionHandler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (gareth-ellis) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsRejectedExecutionHandler.java]: I've had a go at resolving this, though the class is instantiated through EsAbortPolicy and ForceQueuePolicy - the result is needing to pass around the MetricRegistry through quite a few different areas of the code - Is this the way you were meaning? (I havent pushed those changes yet, I can perhaps push to a different branch if you want to take a look, or perhaps I misunderstood)

**Comment by gareth-ellis** [comment]
@elasticmachine update branch

**Comment by stu-elastic** [comment]
@piergm @pgomulka 
What are the next steps for this PR?

**Review by piergm — APPROVED** [review]
LGTM! Thanks @gareth-ellis :D

**Comment by gareth-ellis** [comment]
@elasticmachine update branch

**Comment by gareth-ellis** [comment]
@elasticmachine update branch

**Comment by pgomulka** [comment]
@elasticmachine update branch

File: server/src/internalClusterTest/java/org/elasticsearch/threadpool/SimpleThreadPoolIT.java [hunk 15]
```diff
@@ -95,4 +111,66 @@ public void testThreadNames() throws Exception {
...
+                    .endObject()
+            );
+        }
+        indexRandom(true, builders);
+        int numSearches = randomIntBetween(2, 100);
+        for (int i = 0; i < numSearches; i++) {
+            assertNoFailures(prepareSearch("idx").setQuery(QueryBuilders.termQuery("str_value", "s" + i)));
+            assertNoFailures(prepareSearch("idx").setQuery(QueryBuilders.termQuery("l_value", i)));
+        }
+        plugin.collect();
```
Comment (pgomulka) [inline server/src/internalClusterTest/java/org/elasticsearch/threadpool/SimpleThreadPoolIT.java:144]: @stu-elastic this is how we should trigger the collection right?
I was surprised we did not have this method before, perhaps other metrics were not testing this?

**Review by pgomulka — APPROVED** [review]
LGTM, great work @gareth-ellis !
I left one question for @stu-elastic but I think this is safe for merging
let's not forget to update the PR description (and the commit) as we slightly diverged from what was initially described.
